coordinator/src/main/java/com/tencent/rss/coordinator/PartitionBalanceAssignmentStrategy.java

```diff
@@ -0,0 +1,145 @@
+
+  public PartitionBalanceAssignmentStrategy(ClusterManager clusterManager) {
+    this.clusterManager = clusterManager;
+  }
+
+  @Override
+  public PartitionRangeAssignment assign(
+      int totalPartitionNum,
+      int partitionNumPerRange,
+      int replica,
+      Set<String> requiredTags) {
+
+    if (partitionNumPerRange != 1) {
+      throw new RuntimeException("PartitionNumPerRange must be one");
+    }
+
+    SortedMap<PartitionRange, List<ServerNode>> assignments = new TreeMap<>();
+    synchronized (this) {
+        List<ServerNode> nodes = clusterManager.getServerList(requiredTags);
+        Map<ServerNode, PartitionInfo> newPartitionInfos = Maps.newConcurrentMap();
+        for (ServerNode node : nodes) {
+          PartitionInfo partitionInfo;
+          if (serverToPartitions.containsKey(node)) {
+            partitionInfo = serverToPartitions.get(node);
+            if (partitionInfo.getTimestamp() < node.getTimestamp()) {
+              partitionInfo.resetPartitionNum();
+              partitionInfo.setTimestamp(node.getTimestamp());
+            }
+          } else {
+            partitionInfo = new PartitionInfo();
+          }
+          newPartitionInfos.putIfAbsent(node, partitionInfo);
+        }
+        serverToPartitions = newPartitionInfos;
+        int assignPartitions = totalPartitionNum * replica / clusterManager.getShuffleNodesMax();
+        for (ServerNode node : nodes) {
+          int p = serverToPartitions.get(node).getPartitionNum();
+        }
+        nodes.sort(new Comparator<ServerNode>() {
+          @Override
+          public int compare(ServerNode o1, ServerNode o2) {
+            PartitionInfo partitionInfo1 = serverToPartitions.get(o1);
+            PartitionInfo partitionInfo2 = serverToPartitions.get(o2);
+            double v1 = o1.getAvailableMemory() * 1.0 / (partitionInfo1.getPartitionNum() + assignPartitions);
+            double v2 = o2.getAvailableMemory() * 1.0 / (partitionInfo2.getPartitionNum() + assignPartitions);
+            return -Double.compare(v1, v2);
+          }
+        });
+        List<ServerNode> assignedNodes = nodes.subList(0, clusterManager.getShuffleNodesMax());
+        int idx = 0;
+        for (int partition = 0; partition < totalPartitionNum; partition++) {
+          List<ServerNode> rangeNodes = Lists.newArrayList();
+          for (int rc = 0; rc < replica; rc++) {
+            idx = nextIdx(idx, assignedNodes.size());
+            ServerNode node = assignedNodes.get(idx);
+            serverToPartitions.get(node).incrementPartitionNum();;
+            rangeNodes.add(node);
+          }
+          assignments.put(new PartitionRange(partition, partition), rangeNodes);
+        }
+    }
+    return new PartitionRangeAssignment(assignments);
+  }
+
+  @VisibleForTesting
+  Map<ServerNode, PartitionInfo> getServerToPartitions() {
+    return serverToPartitions;
+  }
+
+  @VisibleForTesting
+  int nextIdx(int idx, int size) {
```

Comment:
duplicated code with BasicAssignmentStrategy

Reply:
I move nextId to RssUtils